lib/types.nix: implement getSubOptions for either

lilyball opened the pull request:
This implements `getSubOptions`, `getSubModules`, and `substSubModules` for the `either` type. This allows us to generate documentation for a bunch of options that were getting missed before.

Because `either` is the key to making recursive types, I had to be careful about the implementation here to avoid breaking on a type like `(pkgs.formats.json {}).type`. To that end, this implementation only looks at child types if they're `submodule` or another `either`. This does mean types like `either str (listOf submodule)` will still be missed, but I'm not sure how else to handle this besides recursively searching all nested types to see if any are equal (`==`) to the current type, and I don't like that solution. I also considered simply stopping on `attrsOf` and `listOf` and allowing other children (this would allow e.g. `either str (nullOr submodule)`) but I thought that was sufficiently fragile. And I considered looking for any recursively-defined type in nixpkgs and overriding `getSubOptions`/`getSubModules` explicitly, but that would mean types defined outside of nixpkgs could still be broken by this.

For the case of `either submodule submodule` I decided that `getSubOptions` should merge all returned options, but `getSubModules` and `substSubModules` should ensure only one child actually has submodules. I don't expect this to matter because a type like that isn't really going to work, this logic makes more sense for something like `either submodule (attrsOf submodule)` except for the fact that we won't recurse into the `attrsOf` (to avoid problems on recursive types).

With this change, I found a number of options whose documentation were missing or broken. Fixes for each of those modules are included as separate commits. Most of them were pretty straightforward, but the `tor` module had a lot of generated options and it wasn't always clear what the best way to handle it was.

I tested this with `nix-build nixos/release.nix -A manual.aarch64-linux`, I'm not sure if there's any other documentation set that I should have tested too. I also didn't touch the lib tests, I'm not sure if this change should have an explicit test written for it, but I did make sure `lib/tests/modules.sh` passes. I'm also not sure if this change warrants a release note.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

lilyball committed:
nixos/dconf: fix option examples

The `example` key of a couple of options didn't evaluate, and this went
unnoticed as this option wasn't part of generated documentation before.

lilyball committed:
nixos/rspamd: fix option documentation

lilyball committed:
nixos/pay-respects: fix option documentation

The `default` references a config value from another module, so it needs
`defaultText` as well.

lilyball committed:
nixos/movim: add missing option documentation

lilyball committed:
lib/types.nix: implement getSubOptions for either

Also implement getSubModules and substSubModules. Because either is key
to any recursively-defined type (such as `(pkgs.formats.json {}).type`),
we can't just blindly recurse into both sides or we'll hit infinite
recursion on a recursive type. To that end, this implementation only
recurses into a child type if that type is submodule or another either.
This means we'll generate documentation for `either str submodule` but
not for `either str (attrsOf submodule)`.

This change allows us to generate documentation for a bunch of options
that are currently being missed.

lilyball committed:
nixos/tor: add missing option documentation

lilyball commented on nixos/modules/services/security/tor.nix:
```diff
@@ -205,6 +212,7 @@ let
   optionFlags = lib.mkOption {
     type = with lib.types; listOf str;
     default = [ ];
+    internal = true;
```
I'm not entirely sure about this one, from my reading of the module it looks like `flags` isn't supposed to be set by users directly, it's calculated from the per-flag attributes.

lilyball commented:
A few thoughts on this PR:

- I could split this up into several separate PRs, since the option documentation fixes don't depend on the `either` change (they just don't affect anything without it)
- Because this change really only affects documentation (and `nixos-option`) maybe it is safe enough to go ahead and do the naive recursion and just fix nixpkgs's recursive types, since any third parties that define their own recursive types will only be affected if they also build documentation (though they would still hit the issue if they use `nixos-option` to inspect any option using their recursive type). I haven't tested to see if this would pick up any more options though, so maybe I should find that out.
- `home-manager` and `nix-darwin` might want to test their own documentation against this PR to see if they have any options that need documentation fixes.

lilyball commented:
I tested out the "make `either` fully recursive" option and while it does add more options, they're entirely within `services.tor`, such as `services.tor.settings.ControlPort.*.GroupWritable`, and they're all ultimately of type `oneOf [ ... (listOf (oneOf [ ... submodule ]))]`. The downside is I had to update 12 recursive types in `formats` plus another 14 modules that declared their own custom recursive types. That's a lot more recursive types than I expected to find.

Alternatively this also suggests we could just update the 2 affected types in `services.tor` (by overriding `getSubOptions`/`getSubModules`/`substSubModules` to defer directly to the submodule) to get the same effect without making `either` fully recursive. There may be options in home-manager or nix-darwin that would also benefit from the fully-recursive either, but the fact that in all of nixpkgs it's just 2 of the types in `services.tor` that are affected suggests that a construct like `either (listOf submodule) …` (where the submodule isn't also at the top level of the either) is sufficiently rare that it may not show up anywhere else at all.

lilyball commented:
Another thing to consider here is `either module (listOf module)` such as `services.tor.settings.TransPort`, which really should generate option documentation for both the direct sub-options and the listOf sub-options, e.g. both `services.tor.settings.TransPort.IsolateClientAddr` and `services.tor.settings.TransPort.*.IsolateClientAddr`. But right now this can't happen because even though `listOf` adds `"*"` to the `prefix`, this isn't reflected in the returned attrset, so there's no way to merge the sub-options from the `module` and the `listOf module`, unless I want to change `either` to return something like `{ left = t1.getSubOptions prefix; right = t2.getSubOptions prefix; }` and that has negative consequences for `nixos-option`.

With that context, we could consider changing `listOf` and `attrsOf` to put the placeholder into the returned attrset, e.g. having `listOf` return `{ "*" = elemType.getSubOptions (prefix ++ [ "*" ]); }`. That way this can be merged with the non-list version of the same module. The downside is `nixos-option` would have to be updated to expect this.

lilyball commented:
The changes to `either` I just pushed are:

- `getSubOptions` now uses `lib.recursiveUpdateUntil` so `either moduleA moduleB` where the two submodules define distinct options nested within the same key will work, e.g. `either (submodule [{options.foo.bar = mkOption { … }; }]) (submodule [{options.foo.qux = mkOption { … }; }])`
- `getSubOptions` now resolves conflicts in favor of the left type if both types define options
- `getSubModules` now returns the left type's submodules if both types define submodules (this affects `services.postgrey`)
- `substSubModules` similarly now substitutes the left type's submodules if both types define submodules

hsjobeki reviewed (changes requested):
Could you split the changes in types.nix and non-lib related things into a speerate PR? I.e. the fixes on some modules should be seperated. This is a common practise for us to avoid mixing lib changes with nixos fixes. I'll spend a bit more time these days looking at this. I fear that the changes in nixos might be due to a breaking change in the either type. Which me must preserve for downstream users. Moving the lib change out of this PR would make it more clear to us.

lilyball commented:
> I fear that the changes in nixos might be due to a breaking change in the either type.

To clarify on this point, the changes in nixos are to fix code that is currently broken but that brokenness wasn't noticed because the options aren't rendered into documentation until the `lib` change. Some of it is documentation where the nix code is simply wrong (e.g. the `dconf` module said `example = literalExpression ''…''` without importing that, so the fix is to say `example = lib.literalExpression ''…''`), some of it fixes the documentation to conform to the restrictions in place when building the manual (`pay-respects` used a deprecated function and had a `default` key that referenced something from another module without using `defaultText`, `rspamd` also did the latter), and some was just options that were missing descriptions (`movim` and `tor`).

All of this stuff could be detected by using the repl to poke around at the module structure, but it went unnoticed because none of these options ended up in the documentation manual before. So technically this `either` change could be a breaking change just by exposing buggy code to documentation rendering, but getting those options in the rendered documentation is the whole point of this change.

hsjobeki commented:
After looking into this for a short bit i am not sure if implementing `getSubOptions` for either is the right approach to solve the documentation problem. `getSubOptions` is only meant to documentation to automatically traverse options and types via `optionAttrSetToDocList´ with either (and derived types such as oneOf) it is possible to create recursive types. 

In case of those reursive types there is no formally correct way to stop / break on a certain point. Thats why we didn't implement `getSubOptions` historically i think.

Even before this PR either was already broken on recursive types - take this simple recursive type for example:

```nix
  simpleJson = types.oneOf [
    types.str
    (types.attrsOf simpleJson)
  ];

#... a module 
{
  options.foo = mkOption {
    type = simpleJson;
  };
}
```

This would fail to produce docs, because either has a recursive description, which would yield an infinitely long string in this case.

The actual JSON type solves this by overriding the description (this is also not ideal):

```nix
          valueType =
            nullOr (oneOf [
              #...
            ])
            // {
              description = "JSON value";
            };
```

We could maybe think about other solutions such as providing documentation hints, instead of trying to smartly recurse an infinite structure.

Some ideas how to solve this problem: Add something like a `docsModel` / `docsType` which requires to be finite and is provided through the interface while creating the type?

hsjobeki reviewed (commented):
After looking into this for a short bit i am not sure if implementing `getSubOptions` for either is the right approach to solve the documentation problem. `getSubOptions` is only meant to documentation to automatically traverse options and types via `optionAttrSetToDocList´ with either (and derived types such as oneOf) it is possible to create recursive types. 

In case of those reursive types there is no formally correct way to stop / break on a certain point. Thats why we didn't implement `getSubOptions` historically i think.

Even before this PR either was already broken on recursive types - take this simple recursive type for example:

```nix
  simpleJson = types.oneOf [  # renders into a tree of either
    types.str
    (types.attrsOf simpleJson)
    (types.listOf simpleJson)
  ];

#... a module 
{
  options.foo = mkOption {
    type = simpleJson;
  };
}
```

This would fail to produce docs, because either has a recursive description, which would yield an infinitely long string in this case.

The actual JSON type solves this by overriding the description (this is also not ideal):

```nix
          valueType =
            nullOr (oneOf [
              #...
            ])
            // {
              description = "JSON value";
            };
```

We could maybe think about other solutions such as providing documentation hints, instead of trying to smartly recurse an infinite structure.

Some ideas how to solve this problem: Add something like a `docsModel` / `docsType` which requires to be finite and is provided through the interface while creating the type?

lilyball commented:
@hsjobeki 

>  Even before this PR either was already broken on recursive types

For the description, yeah, which is why any recursive types have to override that. nixpkgs already has a bunch of definitions of recursive types and they all override the description. I don't really see any alternative to this for recursive types, you _must_ provide a description no matter how you construct it because any automatically-constructed description will be infinite.

And I'm not sure why you said "Even before this PR", as this PR doesn't introduce any new brokenness. It restricts the `getSubOptions` implementation to recurse only when it's provably safe to do so; any infinite type that's actually usable in practice requires some sort of container type, such as a `listOf` or `attrsOf`, and so we can have `either` recurse as long as it stops on one of those. But since we can't actually say "stop on those two", we can only control the next step after the `either`, that's why I went with the current implementation of recursing only into another `either` (which means we're still controlling the recursion) or into `submodule`. And you cannot construct a recursive type with just `either` and `submodule` without also requiring an infinitely-sized value (and recursing infinitely to check that value's type), so that's fine.

> Some ideas how to solve this problem: Add something like a `docsModel` / `docsType` which requires to be finite and is provided through the interface while creating the type?

Can you please explain what you mean by this? Please bear in mind the actual thing this PR aims to solve is submodules that are currently being omitted from documentation because they're part of _finite_ types that use `either`. I did consider the idea that `either` should just unconditionally recurse and that any recursively-defined types need to break that recursion themselves, and it turns out that within nixpkgs alone there are 26 such types that all need to do something like

```diff
       type =
         let
           valueType =
             nullOr (oneOf [
               bool
               int
               float
               str
               path
               (attrsOf valueType)
               (listOf valueType)
             ])
             // {
               description = "JSON value";
+              getSubOptions = prefix: {};
+              getSubModules = null;
+              substSubModules = m: null;
             };
         in
         valueType;
```

That's a lot of recursive types to update just to make `either` recursive, plus any types defined outside of nixpkgs would break too.

The other option here is to say that any type that uses `either` and has submodules needs to instead use some other variant, some `either'` that does recurse, but that approach means we'll continue to have a lot of undocumented options as people just don't realize their submodules aren't getting documented properly.

I have considered introducing an unconditionally-recursive `either'` type just to fix the remaining options in the tor module that are still missing with this PR, but I haven't actually proposed that because I feel like if the tor module is the only module that actually has this problem, perhaps the tor module needs to have its own scoped solution (e.g. having the tor module define such an `either'` type itself).

hsjobeki commented:
What i meant with `docsModel / docsType`.

Something along these lines:

```nix
# New type, or other solution that allows for user defined and formally correct abort criteria
recursiveJson = types.namedUnion "json" [
  types.str
  types.number
  (types.attrsOf (types.typePlug recursiveJson))  # Explicit recursion marker; docs generation returns the name, instead of the type
  (types.listOf (types.typePlug recursiveJson))   # Explicit recursion marker
];
```

This is just an idea; and is probably the other extreme, forcing Users to control their own recursive types, because the cannot safely be traversed.

For this PR, i would probably expect to add some tests, to ensure we don't cause downstream breakages.
Before going ahead i'd still like to ask @infinisil  or @roberth as they are working much longer with the module system than me about their opinion on this pr.

lilyball commented:
> How does it stop traversing and at what point?

The logic here only controls one step of the traversal. So the answer is that it will stop traversing as soon as it hits a child of `either` that is not itself `submodule` or `either`.

In your example `moduleOrSelf` type, it's true that it wouldn't fail to generate docs prior to this PR, but the type itself is already broken, because what it will do is recurse infinitely the moment you assign something other than an attrset, function, or path to the option. If you have

```nix
{
    options.someOpt = mkOption {
        type = moduleOrSelf;
    };
    config.someOpt = [ "foo" ];
}
```

This will end up recursing infinitely as the `check` function for the first nested `submodule` type fails, and so it falls back to the second type, which recurses and attempts the `submodule` again, etc. So the fact that this change will cause documentation generation to fail on this type doesn't matter, since the type is already broken.

> because we can't know what types are composed using `either` in the wider ecosystem

This argument here is precisely why I went with the limited approach of only recursing when it's provably safe to do so, instead of unconditionally recursing and simply fixing all 26 infinite types in nixpkgs. Limiting the recursion to nested `either` and `submodule` types is provably safe because you cannot build an infinite type out of just those two without already being broken. As I said before, any infinite type that's actually usable and correct requires some container type in the infinite recursion (e.g. a list or attrset).

> How does it handle `either submoduleA submoduleB` (with overlapping / non-overlapping options)

I believe the answer is that the option definition from `submoduleA` will win, though it's been a while at this point since I worked through that logic. This PR uses `recursiveUpdateUntil` to merge the subOptions from both types and it orders the update such that the first type "wins" conflicts.

> We can try to push this idea forward if you added some more tests, that ensure this is a non-breaking change.

What sort of tests are you envisioning?

---

Something I did talk about in an [earlier comment](#issuecomment-3044516586) is that if you have a type that looks like

```nix
types.either (submodule {
  options.bar = mkOption {
    type = types.str;
    description = "A string option in the submodule.";
  };
}) (listOf (submodule {
  options.bar = mkOption {
    type = types.str;
    description = "A string option in the listOf submodule.";
  };
}))
```

then we should be able to generate docs for both of these, but we'll actually only generate them for one of them, and that's because the `listOf` `getSubOptions` looks like

```nix
getSubOptions = prefix: elemType.getSubOptions (prefix ++ [ "*" ]);
```

If instead we do

```nix
listOf = {
  …
  getSubOptions = prefix: { "*" = elemType.getSubOptions (prefix ++ [ "*" ]); };
};
attrsWith = {
  …
  getSubOptions = prefix: { "<${placeholder}>" = elemType.getSubOptions (prefix ++ [ "<${placeholder}>" ]); };
};
```

Then doing `either submoduleFoo (listOf submoduleFoo)` or `either submoduleFoo (attrsOf submoduleFoo)` will generate option documentations for both versions. I haven't made that change in this PR largely because we'd have to update `nixos-option` to understand this as well, and it might impact any third-party implementations of something like `nixos-option` (e.g. I actually have my own `nixos-option` implementation I wrote years ago and it would be affected by the change I'm proposing). So I'm inclined to make that change as a separate PR after we're done dealing with this one.

lilyball commented:
> On the counterside merging options using `recursiveUpdate` could probably lead to suprises.

I don't merge using `recursiveUpdate`, I merge using `recursiveUpdateUntil` that stops as soon as either side is an option definition. This way if you have something like

```nix
types.either (submodule {
  options.foo.bar.baz = mkOption {
    type = types.str;
    description = "A string option in foo.bar.";
  };
  options.foo.bar.corge = mkOption {
    type = types.int;
    description = "An int option in foo.bar.";
  };
}) (submodule {
  options.foo.bar.qux = mkOption {
    type = types.int;
    description = "An int option in foo.bar.";
  };
  options.foo.bar.corge.grault = mkOption {
    type = types.str;
    description = "A string option in foo.bar.corge.";
  };
})
```

then `getSubOptions []` will return

```nix
{
  foo.bar = {
    baz = …; # option from first submodule
    qux = …; # option from second submodule
    corge = …; # option from first submodule
    # foo.bar.corge.grault is missing since foo.bar.corge took precedence
  };
}
```

MattSturgeon commented:
> with either (and derived types such as oneOf) it is possible to create recursive types.

It is perfectly possible to create recursive types without `either`.

In fact, most types in `pkgs/pkgs-lib/formats.nix` are recursive, they just don't have any sub-options and they override their type `description` with a fixed value, so it doesn't need any further handling.

> In case of those reursive types there is no formally correct way to stop / break on a certain point. Thats why we didn't implement `getSubOptions` historically i think.

As with other recursive types, it is down to whoever is declaring the recursive type or option to handle it correctly, either by overriding `getSubOptions` or setting `visible = "shallow"`.

I believe the only reason `getSubOptions` hasn't been added to `either` is because it is difficult to do so without breaking end-user code that assumes it doesn't need to handle recursion itself.

MattSturgeon reviewed (commented):
> To make it non-breaking i would expect something like this to pass:
>
> […]
>
> Please add some of the cases you'd expect it to support to: `lib/tests/modules.sh`

I agree, test cases that assert how specific recursive and non-recursive type definitions should behave would be extremely helpful; both for iterating on an implementation and for confidence to merge.

Don't forget we can also test that expected failure modes produce the correct error messages.

> 1. Track recursion depth and abort once it reaches some `limit` 
> 2. Keep a doc-index of traversed nodes (by id/name by user), if an id is traversed again, stop and render a doc-reference to the item in the doc-index. 

These sound like reasonable approaches for _detecting_ unhandled recursion, so that we can display a deprecation message (error or warning).

> 3. Introduce a new either type

Introducing a new `either` type would work, and avoids the need for deprecation, detection, or breaking changes. But it would be a massively disruptive transition, and we'd struggle to come up with better names than `either` and `oneOf`.

lilyball commented:
> It is perfectly possible to create recursive types without `either`.

Sure, if you effectively reimplement `either`. Any recursive type that doesn't either use or reimplement `either` is rather useless and possibly broken. For example, `let type = types.attrsOf type;` is recursive without either, but it's pretty useless, the only type allowed is attsets. Or `let type = listOf type;`, where all you can do is nest lists, without any other type. Perhaps you can do something gimmicky with that, but it's not a type that's going to see any practical use.

> Since we cannot know know what type-trees exist, should be considered a breaking change in its current form.

That's why always recursing in `getSubOptions` would be a breaking change. But this PR proposes a limited recursion. Recursing just into `either` or `submodule` should always be safe to do, because if it hits infinite recursion with just that then the type is already broken, as I've explained in previous comments.

> Track recursion depth and abort once it reaches some `limit`

We can't do that without rewriting how `getSubOptions` works. But it's also unnecessary with the current approach. We'd only need to do something like that if we were going to make `either` unconditionally recurse in `getSubOptions`.

> Introduce a new either type

The only reason to do this is to allow for unconditional recursion in this new type. And I am tempted to do that simply to allow the `tor` module to use that new type to fix the documentation on the few options that remain undocumented after this PR. The name I'd use for this is `either'`.

lilyball commented on lib/types.nix:
```diff
@@ -1407,6 +1410,32 @@ let
+          # to allow finitely-sized values for recursively-defined types, so we have to put limits
+          # on how we recurse when looking for submodules. To that end, we'll look for submodule
+          # children, and only recurse into either itself. This means `either str submodule` will
+          # work, and `either str (attrsOf submodule)` won't, but that's better than nothing.
```
If we don't treat `either` specially and put limits on how it can recurse then we break the moment we hit a recursive type, and that's a pretty significant breaking change.

MattSturgeon commented on lib/types.nix:
```diff
@@ -1407,6 +1410,32 @@ let
+          # work, and `either str (attrsOf submodule)` won't, but that's better than nothing.
+          getSubOptions =
+            prefix:
+            lib.recursiveUpdateUntil
```
E.g. for freeformType's sub-options, we nest them in `_freeformOptions`:
https://github.com/NixOS/nixpkgs/blob/35e52de8975d18876c332e6b5923065a005fa31f/lib/types.nix#L1276-L1282

The relevant docs code in `lib.options.optionAttrSetToDocList` uses `lib.collect lib.isOption` to collect all options, then it does this recursively on all options' `type.getSubOptions` unless the option is marked `visible = false` or `visible = "shallow"`.

https://github.com/NixOS/nixpkgs/blob/a033f62db92b52df2f6b88e8a21083df612bc94b/lib/options.nix#L607

MattSturgeon commented:
> > It is perfectly possible to create recursive types without `either`.
> 
> Sure, if you effectively reimplement `either`. Any recursive type that doesn't either use or reimplement `either` is rather useless and possibly broken. For example, `let type = types.attrsOf type;` is recursive without either, but it's pretty useless, the only type allowed is attsets. Or `let type = listOf type;`, where all you can do is nest lists, without any other type. Perhaps you can do something gimmicky with that, but it's not a type that's going to see any practical use.

I can see how you'd think this when dealing with fairly primitive types.

With submodules you can have fairly useful recursive types, where a sub-option or freeformType re-uses an outer type recursively. E.g.: https://github.com/nix-community/nixvim/blob/ecb75f49d10fe2823b0822e4e95e53f80e426742/docs/modules/page.nix#L12-L23

I imagine `attrTag` and other non-trivial types may have useful cases for recursion.

> What would you actually expect for generated documentation for a type like this? Because it looks to me like generating documentation for this type should produce an infinite amount of docs, as it needs to document not just `child`, but `child.child`, and `child.child.child`, and `child.child.child.child`, and `child.child.child.child.child`, etc.

Correct. That example would need the sub-options to be declared with `visible = "shallow"` to be useful.

> This type is not the sort of infinite type we've been talking about so far. This type is not going to hit a recursion loop when evaluating `getSubOptions`, but I would expect documentation generation to end up trying to walk the option tree forever.

Given `getSubOptions` is primarily intended[^1] as a way for documentation to get sub-options (or at least, a representation of the sub-options suitable for documentation), I'm not sure the distinction you're drawing is especially useful. The docs walking the option tree forever _is_ the main infinite recursion scenario I'm concerned about.

[^1]: Note, for example, submodule sub-options are not necessarily the exact same options you see in a "real" configuration eval; they come from a limited eval that only includes modules embedded into to the submodule-type itself. Therefore, they exclude any modules added via option-definitions. You'll notice this subtlety if you try to inspect the sub-options' `definitions`. They also have a potentially different `prefix`, producing a potentially different option `loc`.

> The only reason to do this is to allow for unconditional recursion in this new type. [...] The name I'd use for this is `either'`.

Maybe we do need a new type to have general recursion support. Incidentally, nixvim has been using an overlay that replaces `either` with one that supports recursive sub-options.

I don't think `either'` is a good name, though. Perhaps `eitherRec` or `recursiveEither` better convey its purpose. Or an `eitherWith {}` that takes configuration settings and is used to construct the general-use `either`. We'd also want to consider how this affects `oneOf`.

lilyball commented:
> With submodules you can have fairly useful recursive types, where a sub-option or freeformType re-uses an outer type recursively. E.g.: https://github.com/nix-community/nixvim/blob/ecb75f49d10fe2823b0822e4e95e53f80e426742/docs/modules/page.nix#L12-L23

This is an interesting type. But it's not the type of recursive type I was talking about because it doesn't infinitely recurse in `getSubOptions`, it just has an infinite amount of options to document unless you break option collection. This type does disprove what I said in my other comment about the `recursivePayload` type though, since I claimed there that this sort of infinitely-nested-options isn't useful.

> Given `getSubOptions` is primarily intended as a way for documentation to get sub-options (or at least, a representation of the sub-options suitable for documentation), I'm not sure the distinction you're drawing is especially useful. The docs walking the option tree forever is the main infinite recursion scenario I'm concerned about.

If you write something that produces an infinite tree of options, then it's not unreasonable to expect you to break that infinite series of options with something like `visible = "shallow"`. Producing an infinite sea of options is not very common, but producing an infinite recursive type is something that keeps happening. My concern with this PR is being able to produce documentation for options that are not currently documented today, without causing infinite recursion on infinite types (types which, notably, don't actually contain any submodules).

It's certainly possible that someone has, out of tree, written something that would produce an infinite sea of options except it uses `either`, and so they didn't write `visible = "shallow"` since they weren't getting the options documented anyway, but I'd expect that to be rather rare if it exists at all, and such a type really should have `visible = "shallow"` anyway because without that, the code is expressing the intent that it document an infinite amount of options.

> I don't think `either'` is a good name, though. Perhaps `eitherRec` or `recursiveEither` better convey its purpose.

What exactly do you think the purpose here is? Because the options that aren't being documented today, that this PR aims to produce documentation for, they _aren't recursive_. That's the thing, `either` is breaking documentation on _finite_ types because it doesn't want to recurse infinitely in order to prove that an infinite type doesn't contain a submodule. This PR fixes _most_ of those options, and the remaining ones are still not recursive, they just do something like `either someSubmodule (attrsOf someSubmodule)`. The fact that these types are finite is exactly why I don't want to force users to call some `eitherRec` type to get them documented, why I want `either` to just work. And my suggested `either'` is just "we can't fix `either` enough to work for all finite options, so if you notice your option isn't getting documented, try using `either'` instead" (and within nixpkgs it's just the `tor` submodule that would use this). Ideally we'd flip this around and make `either` work for everything and say "use `either'` for infinite types that don't have submodules", but that would break all out-of-tree infinite types.

For the options that this PR has to fix because their documentation doesn't evaluate correctly, I'd wager the module authors didn't even realize those options weren't being documented. If we move ahead with a plan that requires users to switch to some alternative `either` variant for finite types to make documentation work, then we'll still have a bunch of undocumented options because people won't realize they need it.

hsjobeki commented:
I don't think these two cases are equivalent:

-  `functionThatDoesntExist`
```nix
description = functionThatDoesntExist "...";
```

This is objectively broken code - a bug in users code.
The fact that this PR exposes users to their bug is due to laziness.

- Recursive submodule

```nix
childModule = types.submodule {
  options.child = mkOption {
    type = types.either types.str childModule;
  };
};
```

valid, correct code

- Type-checks all values correctly
- Has a clear semantic meaning (recursive tree structure)
- Works perfectly at runtime
- Is (presumably) used in real-world configs

The only "problem" is that `getSubOptions` can't traverse it. But that's a limitation of the documentation system, not a bug in the type.

Proposed solutions:

For case 2, we have several options as we pointed out earlier:

1. Add cycle/depth tracking to `getSubOptions`
2. Add `types.typePlug` for users to mark recursion points
3. The ideal solution would involve a migration of getSubOptions to something that can render recursive types properly. This is a generic solution for a generic problem: recursive types exist.

We shouldn't call correct types "broken" just because the documentation system doesn't handle them yet.

If we merge this PR right now it is a breaking change, which was promised to be avoided if possible. 
Users of these types have documentation with limitations; after this PR, users of these types would not have documentation at all.

---

@roberth @infinisil  @MattSturgeon: What if we try merging; and revert if anyone raises an issue?
It would be annoying for them though to figure out why their docs are broken and questionable if they find this PR

roberth commented:
> * `either moduleA moduleB`

This is unfortunately not a useful type.
Here's a doc update that explains it:
 - https://github.com/NixOS/nixpkgs/pull/456505

tl;dr it always picks `moduleA` and that's somewhat intentional.

lilyball committed:
nixos/libvirtd-autosnapshot: fix option documentation

Option descriptions for a submodule were referencing config values from
the parent module. This isn't allowed when building the manual, so just
delete those references.

lilyball commented:
Rebased, and added a new commit to account for a new option that broke the manual rules (referencing a config value from outside the module).

hsjobeki commented:
> A recursion limit just lets us proceed with documentation generation in the face of such a submodule, but it doesn't actually fix that type, and we'll end up generating documentation for N levels of recursive options, which isn't good. 

That was not my original idea. I wanted to abort with a clear error message, that explains how to fix the problem. Rather than having the native "infinite recursion" error. 

Thats the least we should do. Because docs will break downstream. 

Not sure if should do more, as you pointed out it might be "rare" breakages. 
Any foreseeable breakage however should contain actionable messages if feasible.
